### PR TITLE
Add spdlog

### PR DESCRIPTION
**Pull request by chivay**
This change is [Reviewable](https://reviewable.io/reviews/cert-polska/ursadb/42)

### CODE REVIEW

File: libursa/Database.cpp [hunk 1]
```diff
@@ -123,20 +124,20 @@ void Database::save() {
...
-    std::cout << "loaded new dataset " << ds << std::endl;
+    spdlog::info("Loaded new dataset {}", ds);
 }
 
 void Database::update_iterator(const DatabaseName &name, uint64_t byte_offset,
                                uint64_t file_offset) {
     auto it = iterators.find(name.get_id());
     if (it == iterators.end()) {
-        std::cout << "tried to update nonexistent iterator?";
+        spdlog::warn("Tried to pudate nonexistent iterator");
```
Comment (msm-code) [inline libursa/Database.cpp:140]: `pudate` :thinking:

File: libursa/Database.cpp [hunk 2]
```diff
@@ -203,10 +204,8 @@ void Database::commit_task(uint64_t task_id) {
     Task *task = get_task(task_id);
 
     for (const auto &change : task->changes) {
-        std::cout << "change: " << db_change_to_string(change.type) << " "
-                  << change.obj_name << "(" << change.parameter << ")"
-                  << std::endl;
-
+        spdlog::info("Change: {} {}({})", db_change_to_string(change.type),
```
Comment (msm-code) [inline libursa/Database.cpp:207]: Add a space before `(` (it should've been there from the start)

File: libursa/Indexer.cpp [hunk 3]
```diff
@@ -26,12 +27,11 @@ void Indexer::index(const std::string &target) {
 
         builder->index(target);
     } catch (empty_file_error &e) {
-        std::cout << "empty file (skip): " << target << std::endl;
+        spdlog::warn("Empty file (skip): {}", target);
```
Comment (msm-code) [inline libursa/Indexer.cpp:30]: I think this can safely be `info` or even debug? Empty files are legal i guess.

File: libursa/Indexer.cpp [hunk 4]
```diff
@@ -62,7 +62,7 @@ void Indexer::remove_dataset(const OnDiskDataset *dataset_ptr) {
 }
 
 void Indexer::make_spill(DatasetBuilder &builder) {
-    std::cout << "new dataset" << std::endl;
+    spdlog::debug("new dataset");
```
Comment (msm-code) [inline libursa/Indexer.cpp:65]: Do you mind moving this one line lower and also logging `dataset_name`? Just in case

File: libursa/Indexer.cpp [hunk 5]
```diff
@@ -82,7 +82,7 @@ void Indexer::make_spill(DatasetBuilder &builder) {
             }
             register_dataset(merged_name);
         } else {
-            std::cout << "not going to merge" << std::endl;
+            spdlog::debug("Not merging");
```
Comment (msm-code) [inline libursa/Indexer.cpp:85]:
```suggestion
            spdlog::debug("{} datasets, not merging", candidates.size());
```

File: libursa/Indexer.cpp [hunk 6]
```diff
@@ -73,7 +73,7 @@ void Indexer::make_spill(DatasetBuilder &builder) {
             OnDiskDataset::get_compact_candidates(created_dataset_ptrs());
 
         if (candidates.size() >= INDEXER_COMPACT_THRESHOLD) {
-            std::cout << "merge stuff" << std::endl;
+            spdlog::debug("Merging datasets");
```
Comment (msm-code) [inline libursa/Indexer.cpp:76]:
```suggestion
            spdlog::debug("Merging {} datasets", candidates.size());
```

File: libursa/Indexer.cpp (hunk 3 above)
Comment (msm-code) [reply libursa/Indexer.cpp:30]:
```suggestion
        spdlog::debug("Empty file (skip): {}", target);
```

File: libursa/Database.cpp (hunk 2 above)
Comment (msm-code) [reply libursa/Database.cpp:207]:
```suggestion
        spdlog::info("Change: {} {} ({})", db_change_to_string(change.type),
```

File: src/NetworkService.cpp [hunk 7]
```diff
@@ -127,10 +127,10 @@ void NetworkService::handle_dataset_lock_req(WorkerContext *wctx,
             wctx->snap.lock_dataset(ds_name);
         }
 
-        std::cout << "coordinator: dataset locked ok" << std::endl;
+        spdlog::info("coordinator: dataset locked");
```
Comment (msm-code) [inline src/NetworkService.cpp:129]: Maybe move this to the for loop and log every locked dataset (also uppercase)
```
        spdlog::info("Coordinator: dataset {} locked", ds_name);
```

File: src/NetworkService.cpp [hunk 8]
```diff
@@ -127,10 +127,10 @@ void NetworkService::handle_dataset_lock_req(WorkerContext *wctx,
...
         s_send_val<NetLockResp>(backend, NetLockResp::LockOk);
     } else {
-        std::cout << "coordinator: dataset lock denied" << std::endl;
+        spdlog::warn("coordinator: dataset lock denied");
```
Comment (msm-code) [inline src/NetworkService.cpp:132]:
```suggestion
        spdlog::warn("Coordinator: dataset lock denied");
```

File: src/NetworkService.cpp [hunk 9]
```diff
@@ -163,10 +163,10 @@ void NetworkService::handle_iterator_lock_req(WorkerContext *wctx,
     if (!already_locked) {
         wctx->snap.lock_iterator(iterator_name);
 
-        std::cout << "coordinator: iterator locked ok" << std::endl;
+        spdlog::info("coordinator: iterator locked");
```
Comment (msm-code) [inline src/NetworkService.cpp:165]:
```suggestion
        spdlog::info("Coordinator: iterator {} locked", iterator_name);
```

File: src/NetworkService.cpp [hunk 10]
```diff
@@ -163,10 +163,10 @@ void NetworkService::handle_iterator_lock_req(WorkerContext *wctx,
...
         s_send_val<NetLockResp>(backend, NetLockResp::LockOk);
     } else {
-        std::cout << "coordinator: iterator lock denied" << std::endl;
+        spdlog::info("coordinator: iterator lock denied");
```
Comment (msm-code) [inline src/NetworkService.cpp:168]:
```suggestion
        spdlog::info("coordinator: lock denied for iterator {}", iterator_name);
```

File: libursa/Database.cpp (hunk 1 above)
Comment (msm-code) [inline libursa/Database.cpp:140]:
```suggestion
        spdlog::warn("Can't update invalid iterator {}", name.get_id());
```

File: extern/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+target_include_directories(catch INTERFACE .)
+
+add_library(json INTERFACE)
+target_include_directories(json INTERFACE .)
+
+add_library(pegtl INTERFACE)
+target_include_directories(pegtl INTERFACE .)
+
+# target spdlog::spdlog
+add_subdirectory(spdlog)
```
Comment (msm-code) [inline extern/CMakeLists.txt:11]: No newline at the end of file (I know it's nitpicking but it's triggering my OCD)

File: src/Daemon.cpp [hunk 12]
```diff
@@ -201,7 +203,7 @@ Response dispatch_command_safe(const std::string &cmd_str, Task *task,
         Command cmd = parse_command(cmd_str);
         return dispatch_command(cmd, task, snap);
     } catch (std::runtime_error &e) {
-        std::cout << "Command failed: " << e.what() << std::endl;
+        spdlog::error("Command failed: {}", e.what());
```
Comment (msm-code) [inline src/Daemon.cpp:206]: Could use some more debugging info:

```suggestion
        spdlog::error("Task {} failed: {}", task->id, e.what());
```

File: src/Daemon.cpp [hunk 13]
```diff
@@ -215,6 +217,7 @@ int main(int argc, char *argv[]) {
 
     try {
         Database db(argv[1]);
+        spdlog::info("UrsaDB v{}", get_version());
```
Comment (msm-code) [inline src/Daemon.cpp:220]: What about:
```
spdlog::info("                          _ _");
spdlog::info(" _   _ _ __ ___  __ _  __| | |__");
spdlog::info("| | | | '__/ __|/ _` |/ _` | '_ \");
spdlog::info("| |_| | |  \__ \ (_| | (_| | |_) |");
spdlog::info(" \__,_|_|  |___/\__,_|\__,_|_.__/");
```
:thinking:
(plus \\ escapes) 
(not sure though).

File: src/NetworkService.cpp [hunk 14]
```diff
@@ -86,8 +87,7 @@ void NetworkService::run() {
 
 void NetworkService::commit_task(WorkerContext *wctx) {
     uint64_t did_task = wctx->task->id;
-    std::cout << "task " << did_task << ": finished by worker "
-              << wctx->identity << std::endl;
+    spdlog::info("Task {} finished by worker {}", did_task, wctx->identity);
 
     db.commit_task(wctx->task->id);
```
Comment (msm-code) [inline src/NetworkService.cpp:92]:
```suggestion
    db.commit_task(did_task);
```

File: src/Daemon.cpp (hunk 13 above)
Comment (chivay) [reply src/Daemon.cpp:220]: Not a bad idea, although I'll leave it out for now.

File: libursa/DatabaseSnapshot.cpp [hunk 15]
```diff
@@ -178,7 +179,7 @@ void DatabaseSnapshot::reindex_dataset(Task *task,
     task->work_estimated = source->get_file_count() + 1;
 
     source->for_each_filename([&indexer, &task](const std::string &target) {
-        std::cout << "reindexing " << target << std::endl;
+        spdlog::info("Reindexing {}", target);
```
Comment (msm-code) [inline libursa/DatabaseSnapshot.cpp:182]: Why `info` here and `debug` above? Should be the same in both, I suggest `debug`.

**Pull request merged**